README.md
# City of Nodd, the RPG
**NOTE: This game is 18+ as it contains weird content!**

A new roleplaying game set in the City of Nodd. 

Written in HTML, CSS, and JavaScript, it can be played on your browser, no need for plugins or addons!

Changelog.txt
This is the change log for Nodd RPG, detailing additions and changes.

====================
[0.1.4] - 2021-11-28

Additions:
+ Added the shop description of Eggpluck's Discount Familiars and you can examine the creatures. Pets are not buyable yet.
+ Codex additions!
   * Tutorial articles have been added!
   * Bestial Species: Mephitoad, Miggwitch, and Queril.
   * Factions: The six Houses (not yet accessible)
+ Added two more random exploration flair to the Darkling Row. You also can unlock Lehlt and Miggwitch codex entry that way.
   * Scarab crash scene
   * Miggwitch pulling a carriage scene
+ Added metrics settings. Wherever applicable, metrics will be used.
+ Added a new quest where Drusk can give you favour. Visit him when his favour is 15+. You can collect the fruits for him. Reward: 5 free drinks and 75 XP.
+ The following species have been added to starter playable species:
   * Lion (Feline)
   * Tiger (Feline)
   * Jaguar (Feline)
   * Snep (Feline)
   * Leopard (Feline)
   * Cheetah (Feline)
   * Zebra (Equine)
   * Unicorn (Equine)
+ Added two functions to the documentation: player.hasItem and player.itemCount

Changes:
* Tweaked Drusk's appearance so he's properly wearing his Ego Bracer.
* You now need to have at least 5 energy to explore and search the Darkling Row and Viviria District.
* Drinking at Outside Inn now fills bladder by 8 points instead of 5.
* Game settings moved to its own file and it's also more organized, there are now three categories: Gameplay, Interface, and Kinks & Fetishes.
* Sentient Species entries are now sorted alphabetically.

Bug Fixes:
* Fixed a typo at Slyne Sorceress lust spell. The spell is due for a bit of overhaul anyway though.
* Asking Drusk about guards should now unlock Petrids codex entry.
* Starting a new game while there's already a session in progress should now properly reset flags.
* Some fixes to Ego Bracer.
* Some fixes to ticking hourly flags. Should now only trigger each time a new hour starts.
* Fixed the random exploration cooldown flags so now they properly tick down.

====================
[0.1.3] - 2021-11-22
This release adds a few spells, fixes up the tooltips, and a few things to Grud and Drusk.
The Darkling Row is also now a bit livelier.

Additions:
+ Added barebones talk and sex menus for Grud. He's got a pic. Thanks the Noddule ref!
+ Some additions to Drusk:
   * Friendship stat. Starts at 0, caps at 100. Talking and having drinks increase Friendship.
   * Sex menu, scenes are currently barebones. Unlocked at 10 Friendship.
+ Added three spells for testing: Living Salve, Blind, and Rouse. You can purchase the spells from the temporary spell shop.
   * Living Salve heals you. Amount healed is based on your Intelligence.
   * Blind has a chance to blind your opponent. Chance is based on your Intelligence versus your opponent's Willpower.
   * Rouse raises your opponent's arousal. Uses your Intelligence and Corruption versus your opponent's Willpower.
+ Coded more food consumables that will eventually be added. In particular, Sloach. Not accessible yet.
+ Added two random exploration scenes to Darkling Row. One is a silly mode scene, another is a prank from a Nurk.
+ Added blind debuff. Blind halves accuracy.

Changes:
* Clock interface now dynamically changes between black and green depending on time of day.
* Grud's Grub has the menu changed. Tenderlings and Sloach are no longer offered but instead, Grub Bowl. How fitting.
* Increased the gloam cost of recovering mana, and it also now costs 5 energy to recover mana.
* Enabled Special menu in combat. There are no special attacks yet however...
* Refactored the shop code. Black Ditch Courture now uses the refactored code. The goal is less function redundancy.

Bug Fixes:
* Tooltips now properly resize dynamically and the background is also fixed.
* Tweaked the documentation so status effect documentation points to the correct function.

Removed:
- A file tooltip_short.png has been removed as it's no longer needed.

====================
[0.1.2] - 2021-11-16
This release adds Darkling Row exploration scenes! Drusk also gets some talk scenes to add a bit of lore.
Drusk has finally unveiled the menu of beverages for you to try out! But be careful with Nightmare Fuel!

Additions:
+ More Darkling Inn exploration scenes added! Until you discover the Outside Inn, the encounters are predetermined. This should add to the excitement!
+ Outside Inn additions: Drusk talk scenes and drinks menu.
+ Added eye colour!
+ Added sergal appearance description.
+ Added the rest of the codex entries for Sapient Species. Petrid codex entry is now unlockable.

Changes:
* The shops and the Outside Inn are now initially locked.
* Room 6 now requires the key to access. Make sure to talk to Drusk. Currently, it's free, but there will be costs for subsequent nights. No gloam? Suck the bartender off.
* The no-IE page now has buttons that will link you to the corresponding browsers to download.
* Restroom use function is now less placeholdery, with variants depending on whether you have a dick or not. It even accounts for Latex Sheath Suit!
* Moved the player appearance function to its own file due to the size. Should be no visible gameplay changes overall.
* Lust now slowly increases over time, up to a threshold that is dictated by your Libido stat.
* Moved the attribute bars to a tooltip to create room for something new like a map down the line.
* Sergal tails now count as long tails.
* Penile masturbation scene is now much less placeholdery. Thanks Kat!

Bug Fixes:
* Fixed improper nudity description in player appearance screen.

====================
[0.1.1] - 2021-11-13
Notes for developers: Added documentation entry on status effects!

Additions:
+ Attribute selection added to character creation. You can choose your favoured attribute, and your starting attributes.
+ Added Slyne Sorceress as one of the possible random encounters starting at level 3!
+ Added Slyne codex entry.

js/engine/gui.js
// This code holds the positioning of the GUI display. The look of the display is handled through the CSS file.
// This holds some of the most important code for the engine of the game. It turns the buttons on and off, sets their labels and tool tips
// and tells the program which function to go to next.


//Events
document.onmousemove = getMousePosition;
initializeTooltipEvents();

function getMousePosition(event) {
	var tooltipHt = document.getElementById("tooltipMain").style.height;
	tooltipHt = parseInt(tooltipHt.replace("px", ""));
	document.getElementById("tooltipMain").style.top = (event.clientY - (tooltipHt + 20)) + "px";
	document.getElementById("tooltipStats").style.top = (event.clientY) + "px";
    if (event.clientX + 20 < document.documentElement.clientWidth - 420) {
		document.getElementById("tooltipMain").style.left = (event.clientX + 20) + "px";
		document.getElementById("tooltipStats").style.left = (event.clientX + 20) + "px";
	}
    else
        document.getElementById("tooltipMain").style.left = (document.documentElement.clientWidth - 420) + "px";
}

//Stats Pane
function refreshStats() {
	//------------
	// NUMBERS
	//------------
	//Combat Stats
	document.getElementById("hpNum").innerHTML = Math.floor(player.HP) + " / " + player.maxHP();
    document.getElementById("hpNum").title = "HP: " + Math.floor(player.HP) + " / " + player.maxHP();
	document.getElementById("mpNum").innerHTML = Math.floor(player.MP) + " / " + player.maxMP();
    document.getElementById("mpNum").title = "MP: " + Math.floor(player.MP) + " / " + player.maxMP();
	document.getElementById("lustNum").innerHTML = Math.floor(player.lust) + " / " + player.maxLust();
    document.getElementById("lustNum").title = "Lust: " + Math.floor(player.lust) + " / " + player.maxLust() + " \nMinimum: " + player.minLust() + "\nLust Resistance: " + (Math.floor((1 - player.lustVuln) * 1000) / 10) + "%";
	
	//Personal Needs
	document.getElementById("energyNum").innerHTML = Math.floor(player.energy) + " / " + player.maxEnergy();
    document.getElementById("energyNum").title = "Energy: " + Math.floor(player.energy) + " / " + player.maxEnergy();
	document.getElementById("hungerNum").innerHTML = Math.floor(player.hunger) + " / " + player.maxHunger();
    document.getElementById("hungerNum").title = "Satiety: " + Math.floor(player.hunger) + " / " + player.maxHunger();
	document.getElementById("thirstNum").innerHTML = Math.floor(player.thirst) + " / " + player.maxThirst();
    document.getElementById("thirstNum").title = "Hydration: " + Math.floor(player.thirst) + " / " + player.maxThirst();
	document.getElementById("bladderNum").innerHTML = Math.floor(player.bladder) + " / " + player.maxBladder();
    document.getElementById("bladderNum").title = "Bladder: " + Math.floor(player.bladder) + " / " + player.maxBladder();
	
	//Core Stats
	document.getElementById("strNum").innerHTML = Math.floor(player.str);
	document.getElementById("dexNum").innerHTML = Math.floor(player.dex);
	document.getElementById("endNum").innerHTML = Math.floor(player.end);
	document.getElementById("intNum").innerHTML = Math.floor(player.inte);
	document.getElementById("wilNum").innerHTML = Math.floor(player.wil);
	document.getElementById("chaNum").innerHTML = Math.floor(player.cha);
	document.getElementById("libNum").innerHTML = Math.floor(player.lib);
	document.getElementById("corNum").innerHTML = Math.floor(player.cor);
	//Advancement
	document.getElementById("levelNum").innerHTML = player.level;
	document.getElementById("xpNum").innerHTML = player.XP + " / " + (player.getXPNeeded());
	document.getElementById("moneyNum").innerHTML = player.gloam;

	//------------
	// BARS
	//------------
	//Combat Stats
	document.getElementById("hpBar").style.width = Math.floor((player.HP / player.maxHP()) * 100) + "%";
	document.getElementById("mpBar").style.width = Math.floor((player.MP / player.maxMP()) * 100) + "%";
	document.getElementById("lustBar").style.width = Math.floor((player.lust / player.maxLust()) * 100) + "%";
	
	//Personal Needs
	document.getElementById("energyBar").style.width = Math.floor((player.energy / player.maxEnergy()) * 100) + "%";
	document.getElementById("hungerBar").style.width = Math.floor((player.hunger / player.maxHunger()) * 100) + "%";
	document.getElementById("thirstBar").style.width = Math.floor((player.thirst / player.maxThirst()) * 100) + "%";
	document.getElementById("bladderBar").style.width = Math.floor((player.bladder / player.maxBladder()) * 100) + "%";

	//Core Stats
	document.getElementById("strBar").style.width = Math.floor((player.str / player.maxAttribute("strength")) * 100) + "%";
	document.getElementById("dexBar").style.width = Math.floor((player.dex / player.maxAttribute("dexterity")) * 100) + "%";
	document.getElementById("endBar").style.width = Math.floor((player.end / player.maxAttribute("endurance")) * 100) + "%";
	document.getElementById("intBar").style.width = Math.floor((player.inte / player.maxAttribute("intelligence")) * 100) + "%";
	document.getElementById("wilBar").style.width = Math.floor((player.wil / player.maxAttribute("willpower")) * 100) + "%";
	document.getElementById("chaBar").style.width = Math.floor((player.cha / player.maxAttribute("charisma")) * 100) + "%";
	document.getElementById("libBar").style.width = Math.floor((player.lib / 100) * 100) + "%";
	document.getElementById("corBar").style.width = Math.floor((player.cor / 100) * 100) + "%";
	//Advancement
	if ((player.XP / (player.getXPNeeded())) < 1)
		document.getElementById("xpBar").style.width = Math.floor((player.XP / player.getXPNeeded()) * 100) + "%";
	else 
		document.getElementById("xpBar").style.width = "100%";
	//Name
    document.getElementById("charName").innerHTML = player.name;
	refreshTime();
}
function refreshTime() {
	//Time
	var timeText = "";
    timeText += "<b>Date:</b> " + Time.getNoddDate() + "<br><b>Time:</b> ";
    if (use12Hours) {
        if (time.hours < 12) { //am
            if (time.hours == 0)
                timeText += (time.hours + 12) + ":" + (time.minutes < 10 ? "0" : "") + time.minutes;
            else
                timeText += time.hours + ":" + (time.minutes < 10 ? "0" : "") + time.minutes;
            timeText += "am";
        }
        else { //pm
            if (time.hours == 0)
                timeText += time.hours + ":" + (time.minutes < 10 ? "0" : "") + time.minutes;
            else
                timeText += (time.hours - 12) + ":" + (time.minutes < 10 ? "0" : "") + time.minutes;
            timeText += "pm";
        }
    }
    else
        timeText += time.hours + " bells; " + (time.minutes < 10 ? "0" : "") + time.minutes + " chimes";
    document.getElementById("timeDisplay").innerHTML = timeText;
	if (time.hours >= 7 && time.hours < 19) { //Green BG
		document.getElementById("timeDisplay").style.borderImageSource = "url(assets/interface/tooltip.png)";
		document.getElementById("timeDisplay").style.color = "#000000";
	}
	else { //Dark BG
		document.getElementById("timeDisplay").style.borderImageSource = "url(assets/interface/ui_border.png)";
		document.getElementById("timeDisplay").style.color = "";
	}
}
function showStats() {
	refreshStats();
	document.getElementById("stats").style.visibility = "visible";
	document.getElementById("map").style.visibility = "visible";
	document.getElementById("timeDisplay").style.visibility = "visible";
}
function hideStats() {
	document.getElementById("stats").style.visibility = "hidden";
	document.getElementById("map").style.visibility = "hidden";
	document.getElementById("timeDisplay").style.visibility = "hidden";
}

function hideUpDown() {
    var arrows = ["strArrow", "dexArrow", "endArrow", "intArrow", "wilArrow", "chaArrow", "libArrow", "corArrow", "hpArrow", "mpArrow", "lustArrow", "energyArrow", "hungerArrow", "thirstArrow", "bladderArrow"];
    for (var i = 0; i < arrows.length; i++) {
        document.getElementById(arrows[i]).style.visibility = "hidden";
    }
}
function showUpDown(arrowToDisplay, upDown) {
	var invertedArrows = false;
	if (arrowToDisplay == "lustArrow" || arrowToDisplay == "bladderArrow" || arrowToDisplay == "bowelArrow") invertedArrows = true;
    //Auto-route parameter
    if (arrowToDisplay == "inteArrow") arrowToDisplay = "intArrow";
    //Display arrow
    if (upDown == "up")
        document.getElementById(arrowToDisplay).style.backgroundImage = "url(assets/interface/arrow_" + (invertedArrows == true ? "red" : "green") + ".png)";
    else if (upDown == "down")
        document.getElementById(arrowToDisplay).style.backgroundImage = "url(assets/interface/arrow_" + (invertedArrows == true ? "green" : "red") + ".png)";
    document.getElementById(arrowToDisplay).style.visibility = "visible";
}
function displaySprite(spriteId) {
    if (spriteId == undefined) {
        document.getElementById("spriteDisplay").innerHTML = "";
    }
    else {
        var image = new Image();
        image.src = "assets/sprites/" + spriteId + ".png";

        document.getElementById("spriteDisplay").innerHTML = "<img src=\"assets/sprites/" + spriteId + ".png\">"
    }
}

//Bottom menu buttons
function showButtons() {
	for (var i = 0; i < 15; i++) {
		document.getElementById("button" + pos).style.visibility = "visible";
	}
}
function hideButtons() {
	for (var i = 0; i < 15; i++) {
		document.getElementById("button" + pos).style.visibility = "hidden";
	}
}

function showRow(row) {
	document.getElementById("row" + row).style.visibility = "visible";
}
function hideRow(row) {
	document.getElementById("row" + row).style.visibility = "hidden";
}

function menu() {
	for (var i = 0; i < 15; i++) {
		disableButton(i);
	}
}

function genericPlayerButtons(noNavButtons) {
	hideUpDown();
	refreshStats();
	if (!noNavButtons) {
		addButtonDisabled(5, "Up");
		addButtonDisabled(6, "North");
		addButtonDisabled(7, "Down");
		addButtonDisabled(10, "West");
		addButtonDisabled(11, "South");
		addButtonDisabled(12, "East");
	}
	addButtonDisabled(8, "Masturbate", "Not yet implemented anywhere. For now, please visit Room 6 in the upstairs of Outside Inn.");
	addButton(9, "Wait", Time.waitMenu);
	addButton(13, "Inventory", Inventory.inventoryMenu);
	addButtonDisabled(14, "Map", "Not yet implemented.");
}

function addButton(pos, txt, func, arg1, arg2, arg3, tooltipText, tooltipHeader) {
    if (tooltipHeader == undefined) tooltipHeader = txt;
    var callback = createCallBackFunction(func, arg1, arg2, arg3);
	document.getElementById("button" + pos).innerHTML = txt;
    document.getElementById("button" + pos).enabled = true;
	document.getElementById("button" + pos).onclick = callback;
	document.getElementById("button" + pos).style.opacity = "";
    document.getElementById("button" + pos).tooltipHeader = tooltipHeader;
    document.getElementById("button" + pos).tooltipText = tooltipText;
    return document.getElementById("button" + pos);
}
function addButtonDisabled(pos, txt, tooltipText, tooltipHeader) {
	disableButton(pos);
    if (tooltipHeader == undefined) tooltipHeader = txt;
    document.getElementById("button" + pos).innerHTML = txt;
    document.getElementById("button" + pos).tooltipHeader = tooltipHeader;
    document.getElementById("button" + pos).tooltipText = tooltipText;
}
function disableButton(pos) {
	document.getElementById("button" + pos).innerHTML = "";
	document.getElementById("button" + pos).enabled = false;
	document.getElementById("button" + pos).onclick = null;
	document.getElementById("button" + pos).style.opacity = 0.4;
	document.getElementById("button" + pos).tooltipHeader = undefined;
    document.getElementById("button" + pos).tooltipText = undefined;
}

function isButtonEnabled(pos) {
	return document.getElementById("button" + pos).enabled == true;
}

function doNext(func) {
	menu();
	addButton(0, "Next", func);
}
function doYesNo(yesFunc, noFunc) {
	menu();
	addButton(0, "Yes", yesFunc);
	addButton(1, "No", noFunc);
}

//Bottom left menu buttons
function showMenus() {
	document.getElementById("buttonMain").style.visibility = "visible";
	document.getElementById("buttonSettings").style.visibility = "visible";
	document.getElementById("buttonData").style.visibility = "visible";
	document.getElementById("buttonChar").style.visibility = "visible";
}
function hideMenus() {
	document.getElementById("buttonMain").style.visibility = "hidden";
	document.getElementById("buttonSettings").style.visibility = "hidden";
	document.getElementById("buttonData").style.visibility = "hidden";
	document.getElementById("buttonChar").style.visibility = "hidden";
}
function hideMenuButton(menuButton) {
	document.getElementById(menuButton).style.visibility = "hidden";
}
function showMenuButton(menuButton) {
	document.getElementById(menuButton).style.visibility = "visible";
}
function setMenuButton(menuButton, text, func) {
	document.getElementById(menuButton).innerHTML = text;
	document.getElementById(menuButton).onclick = func;
}

//Hook function too
function hint(pos, txt, hdr) {
	if (hdr != undefined) {
		document.getElementById("button" + pos).tooltipHeader = hdr;
	}
	else {
		document.getElementById("button" + pos).tooltipHeader = document.getElementById("button" + pos).textContent;
	}
	document.getElementById("button" + pos).tooltipText = txt;
}

//Tooltip
function initializeTooltipEvents() {
	//Hooks tooltips to buttons.
    for (var i = 0; i < 15; i++) {
		var btn = document.getElementById("button" + i);
        //Create blank variable
        btn.tooltipHeader = undefined;
        btn.tooltipText = undefined;
        //Hook events
        btn.onmouseover = (function(event) {
            if (event.currentTarget.tooltipText != undefined) {
                document.getElementById("tooltipMain").innerHTML = "<h5>" + event.currentTarget.tooltipHeader + "</h5><p>" + event.currentTarget.tooltipText + "</p>";
				var ht = document.getElementById("tooltipMain").childNodes[0].clientHeight + document.getElementById("tooltipMain").childNodes[1].clientHeight + 10;
				document.getElementById("tooltipMain").style.height = ht + "px";
                document.getElementById("tooltipMain").style.visibility = "visible";
            }
        });
        btn.onmouseout = function() {
            document.getElementById("tooltipMain").style.visibility = "hidden";
        }
    }
	//Hooks tooltips to stats.
	var stats = document.getElementById("stats");
	stats.onmouseover = (function(event) {
		document.getElementById("tooltipStats").style.visibility = "visible";
	});
	stats.onmouseout = function() {
		document.getElementById("tooltipStats").style.visibility = "hidden";
	}
}
